File: raat/devices/digital-input.device-plugin/digital-input.cpp
#include "raat.hpp"

#include "digital-input.hpp"

DigitalInput::DigitalInput(uint8_t pin)
{
    m_pin = pin;
}

void DigitalInput::reset() {}
void DigitalInput::tick() {}

void DigitalInput::setup()
{
    pinMode(m_pin, INPUT);
}

bool DigitalInput::state()
{
    return digitalRead(m_pin) == HIGH;
}

uint16_t DigitalInput::command_handler(char const * const command, char * reply)
{
    (void)command;
    reply[0] = digitalRead(m_pin) == HIGH ? '1' : '0';
    reply[1] = '\0';
    return 1;
}

uint8_t DigitalInput::pin()
{
	return m_pin;
}

File: raat/devices/digital-input.device-plugin/digital_input.py
import logging

from pathlib import Path

from collections import namedtuple

from yapsy.IPlugin import IPlugin

from raat.types import DeviceSource, DeviceInclude

from raat.devices.generic_device import GenericDevice

THIS_PATH = Path(__file__).parent


class DigitalInput(GenericDevice, namedtuple("DigitalInput", ["name", "pin"])):

    __slots__ = ()

    sources = (DeviceSource(THIS_PATH, "digital-input.cpp"), )

    includes = (DeviceInclude(THIS_PATH, "digital-input.hpp"), )

    @property
    def setup(self):
        return "{name}.setup();".format(name=self.cname())

    @property
    def directory(self):
        return THIS_PATH

    @property
    def declarations(self):
        return "static DigitalInput {name} = DigitalInput({pin});".format(
            name=self.cname(), pin=self.pin.value)


class DigitalInputPlugin(IPlugin):

    device_class = DigitalInput

    def activate(self):
        pass

    def deactivate(self):
        pass

    def get(self, device):
        return DigitalInput(device.name, device.settings["pin"])

    def set_log_level(self, level):
        logging.getLogger(__name__).setLevel(level)
